### PR TITLE
Feature/vdyp 400 grow qmd+lh

### PR DESCRIPTION
**Pull request by mjunkin**
Implementation complete and unit testing completed up to step 6 of grow().

### CODE REVIEW

**Review by smithkm — CHANGES_REQUESTED** [review]

**Comment by smithkm** [comment]
The comments about magic numbers and unifying some methods in UtilizationOperations are more thoughts I had while reviewing.  They might make more sense to deal with more comprehensively across the code base instead of in this PR.  the DQ magic number is also a bit annoying since there are so many slightly offset versions of it, should we make a constant for each or offset each from the base as it's used?

File: vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessingEngine.java [hunk 1]
```diff
@@ -15,165 +16,2371 @@
...
+					continue;
+				}
+
+				float spDqStart = bank.quadMeanDiameters[i][UC_ALL_INDEX];
+
+				tryDq[i] = 7.5f + (dqs2[i] - 7.5f) * ( (spDqStart - 7.5f) / (dqs1[i] - 7.5f));
+				tryDq[i] = FloatMath.clamp(tryDq[i], dqLowerBoundBySpecies[i], dqUpperBoundBySpecies[i]);
+
+				rs1[i] = FloatMath.log( (spDqStart - 7.5f) / (dqs1[i] - 7.5f));
+				tryTph[i] = BaseAreaTreeDensityDiameter.treesPerHectare(baNew[i], tryDq[i]);
```
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessingEngine.java:820]: Agreed. I've changed uses of the UC constants to lowBound and highBound of the UC enum values. I have not changed comments referring to them, or any of the tests.

File: vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessingEngine.java [hunk 2]
```diff
@@ -15,165 +16,2371 @@
...
+	 */
+	void growLoreyHeights(
+			LayerProcessingState lps, float dhStart, float dhEnd, float pspTphStart, float pspTphEnd, float pspLhStart
+	) {
+		Bank bank = lps.getBank();
+
+		float pspLhStartEstimate = estimatePrimarySpeciesLoreyHeight(dhStart, pspTphStart);
+		float pspLhEndEstimate = estimatePrimarySpeciesLoreyHeight(dhEnd, pspTphEnd);
+
+		float primaryF = (pspLhStart - 1.3f) / (pspLhStartEstimate - 1.3f);
```
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessingEngine.java:1571]: I think it's "breast height" but there are other constants around that value also used. So I'm not really sure how best to handle this and it's out of the scope of these changes anyway, so it'll have to wait.

File: vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessingEngine.java [hunk 3]
```diff
@@ -1078,19 +3274,40 @@ static void determinePolygonRankings(PolygonProcessingState state, Collection<Li
...
-			state.setSpeciesRankingDetails(
+			int basalAreaGroup1 = 0;
+
+			String primarySpeciesName = bank.speciesNames[highestPercentageIndex];
+			String becZoneAlias = bank.getBecZone().getAlias();
+
+			int defaultEquationGroup = fps.fcm.getDefaultEquationGroup().get(primarySpeciesName, becZoneAlias);
+			Optional<Integer> equationModifierGroup = fps.fcm.getEquationModifierGroup()
+					.get(defaultEquationGroup, inventoryTypeGroup);
+			if (equationModifierGroup.isPresent()) {
```
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessingEngine.java:3292]: Agreed...

File: vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/LambdaProcessingException.java
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/LambdaProcessingException.java:1]: I wasn't aware of RuntimeProcessingException. I'll remove Lambda and use it.

File: vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/UtilizationOperations.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	}
+
+	/**
+	 * Implements VDYPGETU lines 224 - 229, in which the utilization-per-hectare values are scaled by the given factor -
+	 * the % coverage of the primary layer.
+	 *
+	 * @param scalingFactor the factor by which the <code>uh</code> is to be scaled
+	 */
+	private static void scale(VdypUtilizationHolder uh, float scalingFactor) {
+
```
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/UtilizationOperations.java:70]: Agreed, added TODO to UtilizationOperations.

File: vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/UtilizationOperations.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+				uh.getCloseUtilizationVolumeNetOfDecayWasteAndBreakageByUtilization()
+						.set(uc, cuVolumeMinusDecayWastageBreakage * scalingFactor);
+			}
+
+			// quadratic mean diameter is not a per-hectare value and therefore not scaled.
+		}
+	}
+
+	private static final float MAX_ACCEPTABLE_BASAL_AREA_ERROR = 0.1f;
+	private static final UtilizationVector CLASS_LOWER_BOUNDS = new UtilizationVector(
```
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/main/java/ca/bc/gov/nrs/vdyp/forward/UtilizationOperations.java:112]: Yes, I momentarily forgot that...

File: vdyp-lib/vdyp-forward/src/test/java/ca/bc/gov/nrs/vdyp/forward/CalculateQuadMeanDiameterDeltaTest.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+		Optional<Float> v_ba_start = Optional.empty();
+		Optional<Float> v_ba_end = Optional.empty();
+		Reference<Boolean> dqGrowthLimitApplied = new Reference<>();
+
+		float gba = fpe.calculateQuadMeanDiameterDelta(
+				yabh, ba, hd, dq, v_ba_start, v_ba_end, growthInHd, dqGrowthLimitApplied
+		);
+
+		assertThat(gba, is(0.3194551f));
+		assertTrue(dqGrowthLimitApplied.isPresent());
```
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/test/java/ca/bc/gov/nrs/vdyp/forward/CalculateQuadMeanDiameterDeltaTest.java:112]: This isn't an Optional.

File: vdyp-lib/vdyp-forward/src/test/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessorTest.java
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/test/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessorTest.java:1]: Seems best to leave it as is for now.

File: vdyp-lib/vdyp-forward/src/test/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessorTest.java [hunk 7]
```diff
@@ -33,8 +40,56 @@ void test() throws IOException, ResourceParseException, ProcessingException {
...
+						);
+						assertTrue(streamEntry.hashCode() == fileEntry.hashCode());
+						assertTrue(streamEntry.getName().equals(fileEntry.getName()));
+					}
+
+					assertFalse(fileEntries.hasNext());
+				}
+			}
+		} finally {
+			Files.delete(zipFilePath);
```
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-forward/src/test/java/ca/bc/gov/nrs/vdyp/forward/ForwardProcessorTest.java]: OK, done.

File: vdyp-lib/vdyp-common/src/test/java/ca/bc/gov/nrs/vdyp/io/ZipOutputFileResolverTest.java
[the remark this answers is not in the document]
Comment (mjunkin) [reply vdyp-lib/vdyp-common/src/test/java/ca/bc/gov/nrs/vdyp/io/ZipOutputFileResolverTest.java:1]: OK, done.

**Review by smithkm — APPROVED** [review]

**Pull request merged**